File: new/python/143_reorder_list_brute_force.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # idea: brute force
        # put the linked list nodes in a list
        # now manipulate the indices with pointers to go zig zag from left end to right end.
        # taking care of the pointers, by increasing on the left end, and decreasing on the right end
        # until left pointer >= right pointer.
        if not head:
            return

        curr = head
        nodes = []
        while curr:
            nodes.append(curr)
            curr = curr.next

        left = 0
        right = len(nodes) - 1

        while left < right:
            nodes[left].next = nodes[right]
            left += 1
            if left >= right:
                break
            nodes[right].next = nodes[left]
            right -= 1

        nodes[left].next = None
        # time: O(n)
        # space: O(n)

File: new/cpp/143_reorder_list_brute_force.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        // idea: brute force
        // put the pointer to each node in the linked list in a dynamic array
        // then match the next pointer of the first node, to the last pointer
        // then last pointers next node to the first node, after first node moves to next place
        // set the last node's next to None (first pointer)
        if (!head) return;

        ListNode* curr = head;
        vector<ListNode*> nodes;

        while (curr) {
            nodes.push_back(curr);
            curr = curr->next;
        }

        int first = 0;
        int last = nodes.size() - 1;

        while (first < last) {
            nodes[first]->next = nodes[last];
            first++;
            if (first >= last) {
                break;
            }
            nodes[last]->next = nodes[first];
            last--;
        }

        nodes[first]->next = nullptr;
    }
    // time: O(n)
    // space: O(n)
};
